OrbitLinuxTracing/Utils.cpp

```diff
@@ -0,0 +1,150 @@
+  int num_cores = static_cast<int>(std::thread::hardware_concurrency());
+  // Some compilers do not support std::thread::hardware_concurrency().
+  if (num_cores != 0) {
+    return num_cores;
+  }
+
+  std::string num_cores_str = ExecuteCommand("nproc");
+  if (!num_cores_str.empty()) {
+    return std::stoi(num_cores_str);
+  }
+
+  return 1;
+}
+
+// Read the cpuset entry in /proc/<pid>/cgroup.
+static std::optional<std::string> GetCgroupCpuset(pid_t pid) {
+  std::string cgroup_filename = absl::StrFormat("/proc/%d/cgroup", pid);
+  std::optional<std::string> cgroup_content_opt = ReadFile(cgroup_filename);
+
+  if (!cgroup_content_opt.has_value()) {
+    return std::optional<std::string>{};
+  }
+  std::istringstream cgroup_content{cgroup_content_opt.value()};
+
+  std::string cgroup_line;
+  while (std::getline(cgroup_content, cgroup_line)) {
+    if (cgroup_line.find("cpuset:") != std::string::npos ||
+        cgroup_line.find("cpuset,") != std::string::npos) {
+      // For example "8:cpuset:/" or "8:cpuset:/game", but potentially also
+      // "5:cpuacct,cpu,cpuset:/daemons".
+      return cgroup_line.substr(cgroup_line.find_last_of(':') + 1);
+    }
+  }
+
+  return std::optional<std::string>{};
+}
+
+static std::optional<std::string> GetCpusetCpusString(
+    const std::string& cgroup_cpuset) {
+  std::string cpuset_cpus_filename =
+      absl::StrFormat("/sys/fs/cgroup/cpuset%s/cpuset.cpus",
+                      cgroup_cpuset == "/" ? "" : cgroup_cpuset);
+  return ReadFile(cpuset_cpus_filename);
+}
+
+static std::vector<int> ParseCpusetCpus(
+    const std::string& cpuset_cpus_content) {
+  std::vector<int> cpuset_cpus{};
+  // Example of format: 0-2,7,12-14
+  for (const auto& range : absl::StrSplit(cpuset_cpus_content, ',')) {
+    std::vector<std::string> values = absl::StrSplit(range, '-');
+    if (values.size() == 1) {
+      int cpu = std::stoi(values[0]);
+      cpuset_cpus.push_back(cpu);
+    } else if (values.size() == 2) {
+      for (int cpu = std::stoi(values[0]); cpu <= std::stoi(values[1]); ++cpu) {
+        cpuset_cpus.push_back(cpu);
+      }
+    }
+  }
+  return cpuset_cpus;
+}
+
+// Read and parse /sys/fs/cgroup/cpuset/<cgroup_cpuset>/cpuset.cpus for the
+// cgroup cpuset of the process with this pid.
+// An empty result indicates an error, as trying to start a process with an
+// empty cpuset fails with message "cgroup change of group failed".
+std::vector<int> GetCpusetCpus(pid_t pid) {
+  // For example "/" or "/game".
+  std::optional<std::string> cgroup_cpuset_opt = GetCgroupCpuset(pid);
+  if (!cgroup_cpuset_opt.has_value()) {
+    return {};
+  }
+
+  // For example "0-2,7,12-14".
+  std::optional<std::string> cpuset_cpus_content_opt =
+      GetCpusetCpusString(cgroup_cpuset_opt.value());
+  if (!cpuset_cpus_content_opt.has_value()) {
+    return {};
+  }
+
+  return ParseCpusetCpus(cpuset_cpus_content_opt.value());
+}
+
+}  // namespace LinuxTracing
+
+#endif  // ORBIT_LINUX_TRACING_UTILS_H_
```

Comment:
Can you please add tests for these?

Reply:
@dimitry- I'll get to it. So I make them non-static and add them to the header?

Reply:
Yes, I think this is small price to pay.

Reply:
#227